qa/L0_trace/test.sh

```diff
@@ -607,11 +611,122 @@ set -e
 kill $SERVER_PID
 wait $SERVER_PID
 
+
+
+# Check `--trace-config` sets arguments properly
+SERVER_ARGS="--trace-config=triton,file=some_file.log --trace-config=level=TIMESTAMPS \
+            --trace-config=rate=4 --trace-config=count=6 --trace-config=mode=triton --model-repository=$MODELSDIR"
+SERVER_LOG="./inference_server_trace_config.log"
+run_server
+if [ "$SERVER_PID" == "0" ]; then
+    echo -e "\n***\n*** Failed to start $SERVER\n***"
+    cat $SERVER_LOG
+    exit 1
+fi
+
+rm -f ./curl.out
+set +e
+code=`curl -s -w %{http_code} -o ./curl.out localhost:8000/v2/models/simple/trace/setting`
+set -e
+if [ "$code" != "200" ]; then
+    cat ./curl.out
+    echo -e "\n***\n*** Test Failed\n***"
+    RET=1
+fi
+if [ `grep -c "\"trace_level\":\[\"TIMESTAMPS\"\]" ./curl.out` != "1" ]; then
+    RET=1
+fi
+if [ `grep -c "\"trace_rate\":\"4\"" ./curl.out` != "1" ]; then
+    RET=1
+fi
+if [ `grep -c "\"trace_count\":\"6\"" ./curl.out` != "1" ]; then
+    RET=1
+fi
+if [ `grep -c "\"log_frequency\":\"0\"" ./curl.out` != "1" ]; then
+    RET=1
+fi
+if [ `grep -c "\"trace_file\":\"some_file.log\"" ./curl.out` != "1" ]; then
+    RET=1
+fi
+
+set -e
+
+kill $SERVER_PID
+wait $SERVER_PID
+
+set +e
+
+# Check opentelemetry trace exporter sends proper info.
+# A helper python script starts listenning on $OTLP_PORT, where
+# OTLP exporter sends traces. It then check that received data contains 
+# expected entries
+SERVER_ARGS="--trace-config=triton,file=some_file.log --trace-config=level=TIMESTAMPS \
+            --trace-config=rate=1 --trace-config=count=6 --trace-config=mode=opentelemetry --trace-config=opentelemetry,url=localhost:$OTLP_PORT --model-repository=$MODELSDIR"
+SERVER_LOG="./inference_server_trace_config.log"
+run_server
+if [ "$SERVER_PID" == "0" ]; then
+    echo -e "\n***\n*** Failed to start $SERVER\n***"
+    cat $SERVER_LOG
+    exit 1
+fi
+
+# This is a simple python code that opens port
+python $TRACE_COLLECTOR $OTLP_PORT $TRACE_COLLECTOR_LOG &
+COLLECTOR_PID=$! 
+
+# To make sure receiver is ready log gets all data
+sleep 3
+
+# Send http request and collect trace
+$SIMPLE_HTTP_CLIENT >> client_update.log 2>&1
+if [ $? -ne 0 ]; then
+    RET=1
+fi
+
+# Send grpc request and collect trace
+$SIMPLE_GRPC_CLIENT >> client_update.log 2>&1
+if [ $? -ne 0 ]; then
+    RET=1
+fi
+# To make sure log gets all data
+sleep 3
```

Comment:
This is very time sensitive  implementation. Our tests must not introduce such behavior. Can you add a comment here to re-design the implementation?
There should be some form of synchronization between tritonclient issuing request and socket_echo_server.

We should ideally have a single python unit test which starts the socket echo in one thread and issue requests in the main thread, then wait for the event with socket echo before checking for the collected trace.

Reply:
Added FIXME in the beginning of the test